### PR TITLE
Add time series forecasting capabilities

### PR DESCRIPTION
**Pull request by mhham**
- Time series forecasting ML task
  - Creation
  - Core settings
  - Algorithms and hyperparameters
  - Guessing

- API node
  - Endpoint creation
  - Endpoint calling

SC-79580

### CODE REVIEW

File: dataikuapi/dss/dataset.py [hunk 1]
```diff
@@ -363,6 +363,29 @@ def uploaded_list_files(self):
     # Don't forget to synchronize with DSSProject.*
     ########################################################
 
+    def create_timeseries_forecasting_ml_task(self, target_variable,
```
Comment (instanceofme) [inline dataikuapi/dss/dataset.py:366]: Ditto

File: dataikuapi/dss/dataset.py [hunk 2]
```diff
@@ -363,6 +363,29 @@ def uploaded_list_files(self):
...
+                                  timeseries_identifiers=None,
+                                  guess_policy="TIMESERIES_DEFAULT",
+                                  wait_guess_complete=True):
+
+        """Creates a new prediction task in a new visual analysis lab
+        for a dataset.
+
+        :param string target_variable: the variable to predict
+        :param string time_variable:  Column to be used as time variable
+        :param list timeseries_identifiers:  List of columns to be used as time series identifiers
```
Comment (instanceofme) [inline dataikuapi/dss/dataset.py:377]: Same remarks as matching arguments in project handle

File: dataikuapi/dss/ml.py [hunk 3]
```diff
@@ -91,6 +92,13 @@ def set_split_explicit(self, train_selection, test_selection, dataset_name=None,
         :param object train_filter: A :class:`~dataikuapi.dss.utils.DSSFilterBuilder` to build the settings of the filter of the train dataset. May be None (won't be changed)
         :param object test_filter: A :class:`~dataikuapi.dss.utils.DSSFilterBuilder` to build the settings of the filter of the test dataset. May be None (won't be changed)
         """
+        if self.mltask_settings["predictionType"] not in {
+            DSSPredictionMLTaskSettings.PredictionTypes.REGRESSION,
+            DSSPredictionMLTaskSettings.PredictionTypes.BINARY,
+            DSSPredictionMLTaskSettings.PredictionTypes.MULTICLASS,
+        }:
+            raise Exception("Explicit splitting is only available for regression, binary, and multiple classification")
```
Comment (instanceofme) [inline dataikuapi/dss/ml.py:100]: Are we sure? (Causal, deep hub…)

**Review by instanceofme — COMMENTED** [review]
First, partial review of the small bits. Missing the bigger piece in `ml.py`

File: dataikuapi/dss/ml.py [hunk 4]
```diff
@@ -1490,13 +1512,471 @@ def __init__(self, raw_settings, hyperparameter_search_params):
...
+    def get_external_feature(self, feature_name):
+        """
+        Gets the feature preprocessing params for a particular external feature. This returns a reference to the
+        feature's settings, not a copy, so changes made to the returned object will be reflected when saving
+
+        :return: A dict of the preprocessing settings for an external feature
+        :rtype: dict
+        """
+        return self.get_feature_preprocessing(feature_name)
+
```
Comment (instanceofme) [inline dataikuapi/dss/ml.py:1781]: Not sure we need a dedicated method for this

File: dataikuapi/dss/ml.py [hunk 5]
```diff
@@ -1490,13 +1512,471 @@ def __init__(self, raw_settings, hyperparameter_search_params):
...
+    def set_timestep_params(self, timeunit=None, n_timeunits=None, end_of_week_day=None):
+        """
+        Sets the time step parameters for the time series forecasting task.
+
+        :param timeunit: time unit for forecasting. Valid values are: MILLISECOND, SECOND, MINUTE, HOUR, DAY,
+                         BUSINESS_DAY, WEEK, MONTH, QUARTER, HALF_YEAR, YEAR
+        :type timeunit: str
+        :param n_timeunits: number of time units within a time step
+        :type n_timeunits: int
+        :param end_of_week_day: only useful for the week time unit. Valid values are: 1 (Sunday), 2 (Monday), ..., 7 (Saturday)
```
Comment (instanceofme) [inline dataikuapi/dss/ml.py:1801]:
```suggestion
        :param end_of_week_day: only useful for the WEEK time unit. Valid values are: 1 (Sunday), 2 (Monday), ..., 7 (Saturday)
```

File: dataikuapi/dss/ml.py [hunk 6]
```diff
@@ -1490,13 +1512,471 @@ def __init__(self, raw_settings, hyperparameter_search_params):
...
+        """
+        Gets the time step parameters for the time series forecasting task. This returns a reference to the
+        time step parameters, not a copy, so changes made to the returned object will be reflected when saving
+
+        :return: A dict of the time step parameters
+        :rtype: dict
+        """
+        return self.mltask_settings["timestepParams"]
+
+    def set_timestep_params(self, timeunit=None, n_timeunits=None, end_of_week_day=None):
```
Comment (instanceofme) [inline dataikuapi/dss/ml.py:1792]: If `get_timestep_params()` returns a dict, maybe this should simply be `set_timestep()`

File: dataikuapi/dss/ml.py [hunk 7]
```diff
@@ -1490,13 +1512,471 @@ def __init__(self, raw_settings, hyperparameter_search_params):
...
+        """
+        Gets the time series resampling parameters for the time series forecasting task. This returns a reference to the
+        time series resampling parameters, not a copy, so changes made to the returned object will be reflected when saving
+
+        :return: A dict of the resampling parameters
+        :rtype: dict
+        """
+        return self.mltask_settings["preprocessing"]["timeseriesSampling"]
+
+    def set_numerical_interpolation_params(self, method=None, constant=None):
```
Comment (instanceofme) [inline dataikuapi/dss/ml.py:1826]: Ditto `set_numerical_interpolation()`, same for others below

File: dataikuapi/dss/ml.py [hunk 8]
```diff
@@ -1490,13 +1512,471 @@ def __init__(self, raw_settings, hyperparameter_search_params):
...
+    def timeseries_identifiers(self, value):
+        """
+        :param value: List of features used as time series identifiers.
+        :type value: list
+        """
+        assert isinstance(value, list)
+        self.mltask_settings["timeseriesIdentifiers"] = value
+
+    @property
+    def quantiles(self):
```
Comment (instanceofme) [inline dataikuapi/dss/ml.py:1958]: Maybe `forecast_quantiles` to avoid ambiguity? (Ditto setter)

File: dataikuapi/dss/ml.py [hunk 9]
```diff
@@ -1490,13 +1512,471 @@ def __init__(self, raw_settings, hyperparameter_search_params):
...
+        """
+        :return: List of quantiles to forecast
+        :rtype: list
+        """
+        return self.mltask_settings["quantilesToForecast"]
+
+    @quantiles.setter
+    def quantiles(self, values):
+        """
+        :param values: List of quantiles to forecast
```
Comment (instanceofme) [inline dataikuapi/dss/ml.py:1968]: Should document 
- 0.5 requirement
- 0.0001-0.9999 inclusive range
- 1e-4 precision

File: dataikuapi/dss/ml.py [hunk 10]
```diff
@@ -1490,13 +1512,471 @@ def __init__(self, raw_settings, hyperparameter_search_params):
...
+        :type values: list
+        """
+        assert isinstance(values, list)
+        assert 0.5 in values, "Quantile 0.5 should always be included"
+        for quantile in values:
+            assert 0.0001 <= quantile <= 0.9999, "Quantile should be between 0.0001 and 0.9999, got {}".format(quantile)
+            assert abs(quantile - math.floor(quantile * 1e4) / 1e4) > 1e-8, "Quantiles should be defined up to 4 digits after decimal point, got {}".format(quantile)
```
Comment (instanceofme) [inline dataikuapi/dss/ml.py:1975]: This seems like it should just be doc

File: dataikuapi/dss/ml.py [hunk 11]
```diff
@@ -1490,13 +1512,471 @@ def __init__(self, raw_settings, hyperparameter_search_params):
...
+        return self.mltask_settings["preprocessing"]["timeseriesSampling"]
+
+    def set_numerical_interpolation_params(self, method=None, constant=None):
+        """
+        Sets the time series resampling numerical interpolation parameters
+
+        :param method: Interpolation method. Valid values are: NEAREST, PREVIOUS, NEXT, LINEAR, QUADRATIC, CUBIC, CONSTANT
+        :type method: str
+        :param constant: number of time units within a time step
+        :type constant: object
```
Comment (instanceofme) [inline dataikuapi/dss/ml.py:1833]: Not sure I get this param
- the name?
- looks like it should be a number but the type is `object`?

(multiple instances)

**Review by instanceofme — CHANGES_REQUESTED** [review]

File: dataikuapi/apinode_client.py [hunk 12]
```diff
@@ -93,6 +93,29 @@ def predict_records(self, endpoint_id, records, forced_generation=None, dispatch
...
+        Forecast using a time series forecasting model on a DSS API node endpoint
+
+        :param str endpoint_id: Identifier of the endpoint to query
+        :param records: List of time series data records to be used as an input for the
+                        time series forecasting model. Each record should be a dict where
+                        keys are features names, and values features values.
+        :param forced_generation: See documentation about multi-version prediction
+        :param dispatch_key: See documentation about multi-version prediction
+
+        :return: a Python dict of the API answer. The answer contains a "results" key (which is an array of result objects)
```
[the remark this answers is not in the document]
Comment (mhham) [reply dataikuapi/apinode_client.py:107]: The main difference is that input records are not aligned with output records for time series forecasting.

We could rename `records` to `past_records` to make it clear that we are using past data to forecast future data.

For the result it is the future/forecasted data, I could write this explicitly in the docstring

File: dataikuapi/dss/ml.py [hunk 13]
```diff
@@ -1543,17 +2023,14 @@ class PredictionTypes:
         MULTICLASS = "MULTICLASS"
 
     def __init__(self, client, project_key, analysis_id, mltask_id, mltask_settings):
-        DSSMLTaskSettings.__init__(self, client, project_key, analysis_id, mltask_id, mltask_settings)
+        AbstractTabularPredictionMLTaskSettings.__init__(self, client, project_key, analysis_id, mltask_id, mltask_settings)
```
[the remark this answers is not in the document]
Comment (mhham) [reply dataikuapi/dss/ml.py:2026]: In the same file `dataikuapi.dss.ml.AbstractTabularPredictionMLTaskSettings`

[the remark this answers is not in the document]
Comment (instanceofme) [reply dataikuapi/dss/ml.py:1789]: I mean

- `AbstractTabularPredictionMLTaskSettings` does not define/override `__init__`
- IIUC (and if I look at the surrounding code), the pattern for this is supposed to be more like the below

```suggestion
        super(DSSPredictionMLTaskSettings, self).__init__(client, project_key, analysis_id, mltask_id, mltask_settings)
```

…prompting the question of why call it this way?
Granted, the former code called `DSSMLTaskSettings.__init__(self, …` but it's an oddity introduced by 4680b60 and it was seemingly not questioned when repeated here.

(ditto for the TS one)

File: dataikuapi/dss/analysis.py [hunk 14]
```diff
@@ -225,6 +225,44 @@ def create_clustering_ml_task(self,
             mltask.wait_guess_complete()
         return mltask
 
+    def create_timeseries_forecasting_ml_task(self, input_dataset, target_variable,
+                                              time_variable=None,
+                                              timeseries_identifiers=None,
+                                              guess_policy="TIMESERIES_DEFAULT",
+                                              wait_guess_complete=True):
+
+        """Creates a new prediction task in a new visual analysis lab for a dataset.
```
Comment (instanceofme) [inline dataikuapi/dss/analysis.py:234]:
```suggestion
        """Creates a new timeseries forecasting task in this visual analysis lab for a dataset.
```

\+ should also fix the "a new visual analysis lab" → "this visual analysis lab" for clustering above

File: dataikuapi/dss/project.py [hunk 15]
```diff
@@ -630,6 +630,45 @@ def create_clustering_ml_task(self, input_dataset,
             mltask.wait_guess_complete()
         return mltask
 
+    def create_timeseries_forecasting_ml_task(self, input_dataset, target_variable,
+                                              time_variable=None,
+                                              timeseries_identifiers=None,
+                                              guess_policy="TIMESERIES_DEFAULT",
+                                              wait_guess_complete=True):
+
+        """Creates a new prediction task in a new visual analysis lab
```
Comment (instanceofme) [inline dataikuapi/dss/project.py:639]:
```suggestion
        """Creates a new timeseries forecasting task in a new visual analysis lab
```

File: dataikuapi/dss/dataset.py [hunk 16]
```diff
@@ -411,6 +411,29 @@ def create_clustering_ml_task(self, input_dataset,
         return self.project.create_clustering_ml_task(self.dataset_name, ml_backend_type=ml_backend_type, guess_policy=guess_policy,
                                                       wait_guess_complete=wait_guess_complete)
 
+    def create_timeseries_forecasting_ml_task(self, target_variable,
+                                              time_variable=None,
+                                              timeseries_identifiers=None,
+                                              guess_policy="TIMESERIES_DEFAULT",
+                                              wait_guess_complete=True):
+
+        """Creates a new prediction task in a new visual analysis lab
```
Comment (instanceofme) [inline dataikuapi/dss/dataset.py:420]:
```suggestion
        """Creates a new timeseries forecasting task in a new visual analysis lab
```

File: dataikuapi/dss/ml.py [hunk 17]
```diff
@@ -1791,6 +1981,235 @@ def get_algorithm_settings(self, algorithm_name):
...
+        :return:
+        """
+        resampling_params = self.get_resampling_params()
+        if method is not None:
+            assert method in self.CATEGORICAL_IMPUTATION_METHODS, "Invalid imputation method, should be in {}".format(self.CATEGORICAL_IMPUTATION_METHODS)
+            resampling_params["categoricalImputeMethod"] = method
+        if constant is not None:
+            resampling_params["categoricalConstantValue"] = constant
+
+    def set_duplicate_timestamp_handling_method(self, method):
```
Comment (instanceofme) [inline dataikuapi/dss/ml.py:2109]: More future-proof if we ever add other params + more consistent with other params

```suggestion
    def set_duplicate_timestamp_handling(self, method):
```

**Review by instanceofme — APPROVED** [review]
Globally LGTM

File: dataikuapi/dss/ml.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (mhham) [reply dataikuapi/dss/ml.py]: Fixed here: https://github.com/dataiku/dataiku-api-client-python/pull/260/commits/37e294de998ba68849847a25eb85c77c940a4d45

**Comment by mhham** [comment]
Moving this PR to dip

**Pull request closed without merging**